packages/integration-sdk-runtime/src/execution/jobState.ts

```diff
@@ -149,14 +150,36 @@ export function createStepJobState({
   uploader,
   onDuplicateEntityKey,
 }: CreateStepJobStateParams): JobState {
+  const hasHasKeyFn = typeof graphObjectStore.hasKey === 'function';
+
   const addEntities = async (entities: Entity[]): Promise<Entity[]> => {
     if (beforeAddEntity) {
       entities = entities.map(beforeAddEntity);
     }
 
+    const set = hasHasKeyFn && entities.length > 1 ? new Set() : undefined;
+
     for (const [index, entity] of entities.entries()) {
       try {
-        duplicateKeyTracker.registerKey(entity._key);
+        if (hasHasKeyFn) {
+          if (graphObjectStore.hasKey!(entity._key)) {
+            throw new IntegrationDuplicateKeyError(
+              `Duplicate _key detected (_key=${entity._key})`,
+            );
+            // if this add entities contains multiple entities we need to verify it doesn't contain any duplicates within
```

Comment:
The logic around this new set is messy. It introduces complexity where the duplicateKeyTracker was very straightforward. Can we move this complexity to a new implementation of the duplicateKeyTracker? 
It could first check the graphObjectStore and then have a temp store that would be exist while in this loop that takes the place of this new set.
It's not a blocker but seems like we could refine the approach?